src/modules/private.ts

```diff
@@ -475,12 +485,37 @@ export default class Private {
     * @signature starkware specific signature for fast-withdrawal
     * }
     */
-  createFastWithdrawal(
+  async createFastWithdrawal(
     params: PartialBy<ApiFastWithdrawal, 'clientId' | 'signature'>,
+    positionId: string,
   ): Promise<{ withdrawal: TransferResponseObject }> {
     const clientId = params.clientId || Math.random().toString().slice(2).replace(/^0+/, '');
     // TODO meet starkware specification
-    const signature = params.signature || Math.random().toString().slice(2).replace(/^0+/, '');
+
+    let signature: string | undefined = params.signature;
+    if (!signature) {
+      if (!this.starkKeyPair) {
+        throw new Error('Fast withdrawal is not signed and client was not initialized with starkPrivateKey');
+      }
+      const fact = this.starkLib.factRegistry.getTransferErc20Fact({
+        recipient: params.toAddress,
+        tokenAddress: this.starkLib.collateralToken.getAddress(),
+        tokenDecimals: this._collateralTokenDecimals,
+        humanAmount: params.creditAmount,
+        salt: nonceFromClientId(clientId),
+      });
+      const conditionalTransfer = new SignableConditionalTransfer({
+        senderPositionId: params.lpPositionId,
+        receiverPositionId: positionId,
+        receiverPublicKey: this.starkKeyPair.publicKey,
+        factRegistryAddress: this.starkLib.factRegistry.getAddress(),
+        fact,
+        humanAmount: params.creditAmount,
```

Comment:
one of these should be using `debitAmount` I'm pretty sure.

AFAIK, the `humanAmount` in the `fact` is the amount of tokens sent on L1 to the user. The `humanAmount` in the `SignableConditionalTransfer`

given that, and the docs here https://github.com/dydxprotocol/proposals/blob/master/proposals/api/v3.MD, I think this second instance should be `params.debitAmount`

Reply:
Right, the signed amount is the L2 transfer and should be debit amount (compare with [librarian](https://github.com/dydxprotocol/librarian/blob/master/src/controllers/api/v3/fast-withdrawals-controller.ts#L185-L193))

The fact amount is the L1 amount and should be credit amount

Reply:
modified